File: python/hashes/0166-fraction-to-recurring-decimal.py
def fractionToDecimal(self, numerator: int, denominator: int) -> str:
    res = '-' if numerator * denominator < 0 else ''
        
    denominator, numerator = abs(denominator), abs(numerator)
    
    if numerator % denominator == 0: 
        return res + str(numerator // denominator)
    
    integ = numerator // denominator
    res += str(integ) + '.'

    numerator = numerator - integ * denominator
    
    rem = numerator % denominator
    hmap = defaultdict(lambda: None)
    frac = ''
    ctr = 0
    while(rem and rem not in hmap):
        hmap[rem] = ctr
        rem *= 10
        frac += str(rem // denominator)
        ctr += 1
        rem %= denominator
    if not rem :
        return res + frac
    else:
        ct = hmap[rem]
        return res + frac[:ct] + '(' + frac[ct:] + ')'

File: cpp/hashes/0166-fraction-to-recurring-decimal.cpp
#include <map>

using namespace std;


string fractionToDecimal(int numerator, int denominator) {
    bool sign = (numerator <= 0 && denominator < 0) || (numerator >= 0 && denominator > 0);
    
    long long num = numerator; 
    long long den = denominator;
    
    num = abs(num); 
    den = abs(den); 
    
    string real = to_string(num / den);
    
    if (!sign) 
        real = '-' + real;
    
    string decimal = ".";
    
    map<int, int> mp;
    mp.clear();
    
    long long rem = num % den;
    
    if (rem == 0) 
        return real;
    
    while (rem != 0 && (mp.find(rem) == mp.end())) {
        mp[rem] = decimal.length();
        rem = rem * 10;
        decimal += (char)('0' + rem / den);
        rem = rem % den;
    }
    
    if (rem == 0)
        return real + decimal;
    
    return real + decimal.substr(0, mp[rem]) + '(' + decimal.substr(mp[rem]) + ')';
}